cachi2/core/package_managers/rubygems/parser.py

```diff
@@ -1,9 +1,12 @@
 import logging
-from typing import Optional, Union
+import re
+from typing import Iterable, Optional, Union
 
 import pydantic
+from gemlock_parser.gemfile_lock import GEM, GIT, PATH, Gem, GemfileLockParser  # type: ignore
```

Comment:
This is a vendored version of https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/packagedcode/gemfile_lock.py which was updated in Feb 2024 whereas our code was updated in 2022.
While it's okay to copy-paste parts of the code instead of reinventing the wheel, when adding new code to this code base we should take the opportunity and evaluate if things can be improved upon given current restrictions, limitations, updates, technologies, etc.

In this particular case, when I look at both the vendored parser and its origin, that's a few hundred lines of code that need to be maintained (in this case, us, and it's not happening). So, I wonder if a we'd not be better off with a significantly shorter snippet of Ruby code checked into this code base, simply because Ruby's `bundler` library contains primitives to parse the lockfile, so we should leverage its capability to do the heavy lifting for us if possible. For example, this snippet of code is all it's supposedly needed to parse the lockfile and dump it as JSON we would then parse in our Python code (diclaimer: I did generate this prototype with an AI as I can't write any Ruby, but it's not rocket science and it seems it should be easily adjusted to our needs):

```
require 'bundler'
require 'json'

# Load and parse the Gemfile.lock using Bundler::LockfileParser
lockfile_contents = File.read('Gemfile.lock')
parser = Bundler::LockfileParser.new(lockfile_contents)

# Prepare a hash structure to hold the parsed data
parsed_data = {}

parser.specs.each do |spec|
  parsed_data[spec.name] = {
    version: spec.version.to_s,
    dependencies: spec.dependencies.map { |dep| { name: dep.name, requirement: dep.requirement.to_s } }
  }
end

# Convert the parsed data to JSON
json_output = JSON.pretty_generate(parsed_data)

# Output the JSON to a file or stdout
File.write('parsed_gemfile_lock.json', json_output)
puts json_output
```

Note this example doesn't fill in the `remote` string, but I think it could be adapted if the parser class is further inspected. The point of this comment is whether we want to rely on a few hundred lines of unmaintained code vs using a few lines of Ruby dealing with it natively. Also note that we're spawning so many subprocess calls already, so an extra call to ruby to execute such a snippet is negligible.

Reply:
interesting...
I slightly updated the script:
```
#!/usr/bin/env ruby

require 'bundler'
require 'json'

if ARGV.empty?
    puts "Usage: #{__FILE__} path/to/Gemfile.lock"
    exit 1
  end
  
  gemfile_lock_path = ARGV[0]
  
  begin
    lockfile_content = File.read(gemfile_lock_path)
  rescue Errno::ENOENT
    puts "Error: File not found - #{gemfile_lock_path}"
    exit 1
  end

parser = Bundler::LockfileParser.new(lockfile_content)

parsed_data = {}

parser.specs.each do |spec|
    source_details = {}
  
    case spec.source
    when Bundler::Source::Rubygems
      source_details[:type] = 'rubygems'
      source_details[:url] = spec.source.remotes.map(&:to_s).first
      source_details[:checksum] = spec.respond_to?(:checksum) ? spec.checksum : nil
    when Bundler::Source::Git
      source_details[:type] = 'git'
      source_details[:url] = spec.source.uri
      source_details[:revision] = spec.source.revision
      source_details[:branch] = spec.source.branch
    when Bundler::Source::Path
      source_details[:type] = 'path'
      source_details[:subpath] = spec.source.path.to_s
    else
      source_details[:type] = 'unknown'
    end
  
    gem_data = {
      version: spec.version.to_s,
      dependencies: spec.dependencies.map { |dep| { name: dep.name, requirement: dep.requirement.to_s } },
      type: source_details[:type]
    }
  
    gem_data[:checksum] = source_details[:checksum] if source_details[:checksum]
    gem_data[:url] = source_details[:url] if source_details[:url]
    gem_data[:revision] = source_details[:revision] if source_details[:revision]
    gem_data[:branch] = source_details[:branch] if source_details[:branch]
    gem_data[:subpath] = source_details[:subpath] if source_details[:subpath]
  
    parsed_data[spec.name] = gem_data
  end

json_output = JSON.pretty_generate(parsed_data)
File.write('parsed-gems.json', json_output)
```

There must be (at least empty) `Gemfile` in the same directory as `Gemfile.lock`.
Usage:
`ruby <script> <path-to-Gemfile.lock>`

Reply:
> There must be (at least empty) Gemfile in the same directory as you run the script.

If so, then cachi2 needs to make sure the repo is correctly structured before proceeding with processing it, for example with such a script.

Reply:
> There must be (at least empty) Gemfile in the same directory as you run the script.

Sounds dangerous, are we sure the ruby library doesn't execute the Gemfile?

Reply:
Here is the problem - https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/shared_helpers.rb#L17

But we can set `BUNDLE_GEMFILE` to something non-empty as a hack :smile: 
e.g. `BUNDLE_GEMFILE="..." ruby <script> <path-to-Gemfile.lock>`

https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/shared_helpers.rb#L248

Reply:
> But we can set BUNDLE_GEMFILE to something non-empty as a hack 😄

@slimreaper35 I'm confused, isn't the presence of both a `Gemfile` and `Gemfile.lock` mandatory otherwise it's a repo misconfiguration? If so, then we need to the usual sequence of checking mandatory files, path validations, symlink checks, etc. before proceeding with parsing, am I missing something?

Reply:
In the user's repo yes, but I have this script locally in cachi2 repo and pointing to some Gemfile.lock.  When I run the script, ruby is complaining that _Could not locate Gemfile_, which apparently can be silenced by setting env variable BUNDLE_GEMFILE to non-empty value.

Reply:
> In the user's repo yes, but I have this script locally in cachi2 repo and pointing to some Gemfile.lock. When I run the script, ruby is complaining that Could not locate Gemfile (in cachi2 repo), which apparently can be silenced by setting env variable BUNDLE_GEMFILE to non-empty value.

This IMO irrelevant to the intended use, note that such a script shouldn't ever be used outside of our own Python code, so one can assume it's only ever going to be executed on a correctly configured repo. Any other use case can be considered invalid and so we don't need any variable settings even if possible in this particular case you mention.

Reply:
> note that such a script shouldn't ever be used outside of our own Python code

I am confused. Where should the script be used ?

Reply:
> Sounds dangerous, are we sure the ruby library doesn't execute the Gemfile?

This is the main thing we should be investigating. Until now, we considered that all the heavy lifting was done by custom Python code. Adding Bundler to the equation changes the fundamentals of our initial design (which is not a problem per se). So we need to make sure that no arbitrary code execution is ever possible. 

Then, give some thought about how to maintain the Ruby/Bundler binaries in our image: is installing a single version enough (should be, right)? Or will we rely on the Bundler "toolchain" mechanism. 

Otherwise, this is a good improvement from our vendored code.

Reply:
> Sounds dangerous, are we sure the ruby library doesn't execute the Gemfile?

@chmeliik unlikely, I went ahead and spent a few hours debugging Ruby and looking at the code. It doesn't interact with the Gemfile in any way. Let me expand.

> Here is the problem - https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/shared_helpers.rb#L17

While this is technically true, it doesn't tell anyone what's happening underneath, here's the full stacktrace for the script @slimreaper35 provided:
```
/usr/local/bundle/gems/bundler-2.5.17/lib/bundler/shared_helpers.rb:17:in `root': Could not locate Gemfile (Bundler::GemfileNotFound)
        from /usr/local/bundle/gems/bundler-2.5.17/lib/bundler.rb:325:in `root'
        from /usr/local/bundle/gems/bundler-2.5.17/lib/bundler.rb:101:in `bundle_path'
        from /usr/local/bundle/gems/bundler-2.5.17/lib/bundler/source/git.rb:244:in `cache_path'
        from /usr/local/bundle/gems/bundler-2.5.17/lib/bundler/source/git.rb:375:in `git_proxy'
        from /usr/local/bundle/gems/bundler-2.5.17/lib/bundler/source/git.rb:253:in `revision'
        from /test.rb:37:in `block in <main>'
        from /test.rb:26:in `each'
        from /test.rb:26:in `<main>'
```
More specifically, bundler is just trying to set the caching dir path where the cloned repos should reside. In order to calculate the default path it's trying to locate `Gemfile` and then just use the parent stem of the path. Here is an ordered list of relevant calls made:

- https://github.com/rubygems/rubygems/blob/3fdc2f627e6bb476b722c7d37b61074430dea82e/bundler/lib/bundler/source/git.rb#L240
- https://github.com/rubygems/rubygems/blob/3fdc2f627e6bb476b722c7d37b61074430dea82e/bundler/lib/bundler.rb#L100
- https://github.com/rubygems/rubygems/blob/3fdc2f627e6bb476b722c7d37b61074430dea82e/bundler/lib/bundler.rb#L324

And that's it, provided explicit revisions are using in the lockfile, not even `git rev-parse` is run which otherwise would, but for handling of the deps there are different modules like `fetcher` and `installer`.

Reply:
> Then, give some thought about how to maintain the Ruby/Bundler binaries in our image: is installing a single version enough (should be, right)? Or will we rely on the Bundler "toolchain" mechanism.

@brunoapimentel in general, we may need to do something like with Yarn and corepack, but it's irrelevant to this particular case we're discussing IMO, because no binary apart from the Ruby interpreter is involved because everything are just standard library calls.

Reply:
> I am confused. Where should the script be used ?

@slimreaper35 you'd use it from the `rubygems` or what exact name ends up being Python module as a `subprocess.call` where you specify the exact PATH context for the call and you should completely control the execution env. Such a subprocess call would only be made once cachi2 verifies the repo configuration.